File: Leetcode/0965/965_Univalued_Binary_Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isUnivalTree = function (root) {
    return traversal(root, root.val);
};

const traversal = (node, val) => {
    if (node) {
        const nodeVal = node.val;
        if (val !== nodeVal) {
            return false;
        }

        return traversal(node.left, val) && traversal(node.right, val);
    }
    return true;
}

File: Leetcode/0965/965_Univalued_Binary_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* prev;
    bool isUnivalTree(TreeNode* root) {
        if(root == nullptr)
        {
            return true;
        }
        
        if(root->left != nullptr)
        {
            if(!isUnivalTree(root->left))
            {
                return false;
            }
            if(root->val != root->left->val)
            {
                return false;
            }
        }
        
        if(root->right != nullptr)
        {
            if(!isUnivalTree(root->right))
            {
                return false;
            }
            if(root->val != root->right->val)
            {
                return false;
            }
        }
        
        return true;
    }
};